source/hpack/decoder.d

```diff
@@ -19,239 +19,102 @@ import std.typecons : tuple;
-	  * :path: /sample/path
-	  */
-	block = [0x04, 0x0c, 0x2f, 0x73, 0x61, 0x6d, 0x70, 0x6c, 0x65, 0x2f, 0x70, 0x61, 0x74, 0x68];
-	decoder.put(block);
-	assert(decoder.front.name == ":path" && decoder.front.value == "/sample/path");
-
-
-	/** 3. Literal header field never indexed (raw)
-	  * password: secret
-	  */
-	block = [0x10, 0x08, 0x70, 0x61, 0x73, 0x73, 0x77, 0x6f, 0x72, 0x64, 0x06, 0x73, 0x65,
-		  0x63, 0x72, 0x65, 0x74];
-	decoder.put(block);
-	assert(decoder.front.name == "password" && decoder.front.value == "secret");
-	assert(decoder.front.neverIndex);
-
+private size_t decodeInteger(I)(ref I src, ubyte bbuf) @safe @nogc
+{
+	uint nbits = 7;
+	auto res = bbuf.toInteger(1);
+
+	if (res < (1 << nbits) - 1) {
+		return res;
+	} else {
+		uint m = 0;
+		do {
+			// take another octet
+			bbuf = src[0];
+			src = src[1..$];
+			// concatenate it to the result
+			res = res + bbuf.toInteger(1)*(1 << m);
+			m += 7;
+		} while(bbuf == 1);
+		return res;
+	}
+}
 
-	/** 4. Indexed header field (integer)
-	  * :method: GET
-	  */
-	import vibe.http.common;
-	block = [0x82];
-	decoder.put(block);
-	assert(decoder.front.name == ":method" && decoder.front.value == HTTPMethod.GET);
+private void decodeLiteral(I,R)(ref I src, ref R dst) @safe
+{
+	ubyte bbuf = src[0];
+	src = src[1..$];
 
-	/** 5. Full request without huffman encoding
-	  * :method: GET
-      * :scheme: http
-      * :path: /
-      * :authority: www.example.com
-      * cache-control: no-cache
-	  */
-	block = [0x82, 0x86, 0x84, 0xbe, 0x58, 0x08, 0x6e, 0x6f, 0x2d, 0x63, 0x61, 0x63, 0x68, 0x65];
-	table.insert(HTTP2HeaderTableField(":authority", "www.example.com"));
-	auto rdec = HeaderDecoder!(ubyte[])(block, table);
-	HTTP2HeaderTableField[] expected = [
-		HTTP2HeaderTableField(":method", HTTPMethod.GET),
-		HTTP2HeaderTableField(":scheme", "http"),
-		HTTP2HeaderTableField(":path", "/"),
-		HTTP2HeaderTableField(":authority", "www.example.com"),
-		HTTP2HeaderTableField("cache-control", "no-cache")];
+	bool huffman = (bbuf & 128) ? true : false;
 
-	foreach(i,h; rdec.enumerate(0)) {
-		assert(h == expected[i]);
-	}
+	assert(!src.empty, "Cannot decode from empty range block");
 
-	/** 5. Full request with huffman encoding
-	  * :method: GET
-      * :scheme: http
-      * :path: /
-      * :authority: www.example.com
-      * cache-control: no-cache
-	  */
-	block = [0x82, 0x86, 0x84, 0xbe, 0x58, 0x86, 0xa8, 0xeb, 0x10, 0x64, 0x9c,0xbf];
-	auto hdec = HeaderDecoder!(ubyte[])(block, table);
+	// take a buffer of remaining octets
+	auto vlen = bbuf.toInteger(1); // value length
+	auto buf = src[0..vlen];
+	src = src[vlen..$];
 
-	foreach(i,h; hdec.enumerate(0)) {
-		assert(h == expected[i]);
+	if(huffman) { // huffman encoded
+		auto adst = appender!string; // TODO a proper allocator
+		adst.reserve(vlen*2); // max compression ratio is < 0.5
+		decodeHuffman(buf, adst);
+		dst = adst.data;
+	} else { // raw encoded
+		dst = cast(string)buf;
```

Comment:
...and this call `dst.put(cast(string)buf);`?

Reply:
Can't you simply use an `appender!string` there? But regardless of that this is an interesting place with respect to `@nogc` - I'm afraid we'll have to defer that, but at least the code is now 90% there.

Reply:
Yes, but since `appender!string` does not support `clear()` that would mean initializing a new appender for each call to `decodeLiteral`. I figured that wasn't as clean as using an allocator which can be passed through by `vibe-http`. I'll start working to include HPACK in `vibe-http` right away.

Reply:
The problem with `clear()` is that it will corrupt all previous strings extracted from that appender, so the cast to `immutable` is actually unsafe in practice in that case. I agree that an external allocator (with `AllocAppender`) would be better.